go-controller/pkg/factory/factory.go

```diff
@@ -223,24 +346,33 @@ func NewWatchFactory(c kubernetes.Interface, stopChan chan struct{}) (*WatchFact
 
 	go func() {
 		<-stopChan
-		wf.shutdown()
-	}()
 
-	return wf, nil
-}
+		// Remove all informer handlers
+		for _, inf := range wf.informers {
+			inf.Lock()
+			for _, handler := range inf.handlers {
+				if err := handler.kill(); err != nil {
+					logrus.Warningf("failed to remove handler: %v", err)
+				}
+				delete(inf.handlers, handler.id)
+			}
+			inf.Unlock()
```

Comment:
I will never *not* complain about bare `Unlock`s. Proper golang/k8s style seems to be to assume that code might randomly panic and screw you if you don't use `defer`. Especially in cases like here where you're calling out to some mystery function (`handler.kill`) that may or may not be safe. Also, don't you want `inf` to still be locked for the closing event channels part below anyway? (Or if not then that needs a comment explaining why.)

It seems like this made more sense as `wf.shutdown()` than just inlining it

Reply:
Fair enough.